src/Tribe/Commerce/PayPal/Gateway.php

```diff
@@ -0,0 +1,272 @@
+
+		if ( $this->identity_token ) {
+			// if there's an identity token set, we handle payment confirmation with PDT
+			tribe( 'tickets.commerce.paypal.handler.pdt' );
+		} else {
+			// if there isn't an identity token set, we use IPN
+			tribe( 'tickets.commerce.paypal.handler.ipn' );
+		}
+	}
+
+	/**
+	 * Set up hooks for the gateway
+	 *
+	 * @since TBD
+	 */
+	public function hook() {
+		add_action( 'template_redirect', array( $this, 'add_to_cart' ) );
+	}
+
+	/**
+	 * Handles adding tickets to cart
+	 *
+	 * @since TBD
+	 */
+	public function add_to_cart() {
+		global $post;
+
+		// bail if this isn't a Tribe Commerce PayPal ticket
+		if (
+			empty( $_POST['product_id'] )
+			|| empty( $_POST['provider'] )
+			|| 'Tribe__Tickets__Commerce__PayPal__Main' !== $_POST['provider']
+		) {
+			return;
+		}
+
+		$url           = $this->get_cart_url( '_cart' );
+		$now           = time();
+		$post_url      = get_permalink( $post );
+		$currency_code = tribe_get_option( 'ticket-paypal-currency-code' );
+
+		// @TODO: set to the success page
+		$notify_url    = add_query_arg( array( 'bacon' => 1 ), get_permalink( $post ) );
+
+		$args = array(
+			'cmd'           => '_cart',
+			'add'           => 1,
+			'business'      => urlencode( tribe_get_option( 'ticket-paypal-email' ) ),
+			'bn'            => 'ModernTribe_SP',
+			'notify_url'    => urlencode( $notify_url ),
+			'shopping_url'  => urlencode( $post_url ),
+			'currency_code' => $currency_code ? $currency_code : 'USD',
+			'custom'        => 'user_id=' . get_current_user_id(),
+		);
+
+		foreach ( $_POST['product_id'] as $ticket_id ) {
+			$ticket   = tribe( 'tickets.commerce.paypal' )->get_ticket( $post->ID, $ticket_id );
+			$quantity = absint( $_POST[ "quantity_{$ticket_id}" ] );
+
+			// skip if the ticket in no longer in stock or is not sellable
+			if (
+				! $ticket->is_in_stock()
+			|| ! $ticket->date_in_range( $now )
+			) {
+				continue;
+			}
+
+			// if the requested amount is greater than remaining, use remaining instead
+			if ( $quantity > $ticket->remaining() ) {
+				$quantity = $ticket->remaining();
+			}
+
+			// enforce purchase limit
+			if ( $ticket->purchase_limit && $quantity > $ticket->purchase_limit ) {
+				$quantity = $ticket->purchase_limit;
+			}
+
+			// if the ticket doesn't have a quantity, skip it
+			if ( empty( $quantity ) ) {
+				continue;
+			}
+
+			$args['quantity']    = $quantity;
+			$args['amount']      = $ticket->price;
+			$args['item_number'] = "{$post->ID}:{$ticket->ID}";
+			$args['item_name']   = urlencode( $this->get_product_name( $ticket, $post ) );
+
+			// we can only submit one product at a time. Bail if we get to here because we have a product
+			// with a requested quantity
+			break;
+		}
+
+		// If there isn't a quantity at all, then there's nothing to purchase. Redirect with an error
+		if ( empty( $args['quantity'] ) ) {
+			// @TODO: add an error for display
+			wp_redirect( $post_url );
```

Comment:
it _seems_ like this redirect will be internal; will remain on the WP site, i mean. is that true? if so, would recommend `wp_safe_redirect()`